blog/content/post/2024-03-15-uefi-sb-implementation-mpl.md

```diff
@@ -0,0 +1,385 @@
+command line or loading unsigned boot components. When Secure Boot is enabled,
+the user can only access the command line if the user authentication is enabled,
+in which case it is protected by a password.
+
+The combination of the mechanisms vastly increases the level of security and
+trustworthiness of the boot process, however in this section we will focus on
+implementing regular UEFI SB.
+
+### Implementation
+
+To integrate the mechanisms available in `meta-secure-core` you need to
+integrate that layer into your build. We use
+[kas-container](https://github.com/siemens/kas/blob/master/kas-container) to set
+up bitbake projects.
+
+In your kas configuration file add the following to the `repos` section:
+
+```yaml
+meta-secure-core:
+url: https://github.com/Wind-River/meta-secure-core.git
+refspec: 8dc9f1b4a735eccee65a8896760e473e110d147e
+layers:
+    meta-efi-secure-boot:
+    meta:
+    meta-signing-key:
+```
+
+Define `meta-secure-core` variables in your layer's `local.conf`:
+
+```conf
+# UEFI Secure Boot variables
+# Use only UEFI Secure Boot without MOK Secure Boot
+UEFI_SELOADER = "0"
+GRUB_SIGN_VERIFY = "1"
+UEFI_SB = "1"
+
+# We want grub-efi from meta-efi-secure-boot to install bootfiles under
+# /boot/EFI/BOOT
+EFI_BOOT_PATH = "/boot/EFI/BOOT"
+
+DISTRO_FEATURES_NATIVE:append = " efi-secure-boot"
+DISTRO_FEATURES:append = " efi-secure-boot modsign"
+MACHINE_FEATURES_NATIVE:append = " efi"
+MACHINE_FEATURES:append = " efi"
+
+DEBUG_FLAGS:forcevariable = ""
+IMAGE_INSTALL:append = " kernel-image-bzimage"
+```
+
+Also, make sure to install the `efi-secure-boot` packagegroup into your image:
+
+```bitbake
+IMAGE_INSTALL:append = " \
+    packagegroup-efi-secure-boot \
+"
+```
+
+We were using a custom version of the `bootimg-efi` wic plugin to set up a
+UEFI-compliant image. That required us to install Secure Boot-related files in
+a specific directory:
+
+```bitbake
+# grub-efi_%.bbappend
+
+do_deploy:append:class-target () {
+    # deploy all UEFI SB related bootfiles to later use them with bootimg-efi
+    # wic plugin
+    install -d ${DEPLOY_DIR_IMAGE}/bootfiles
+    cp -r ${D}/boot/* ${DEPLOY_DIR_IMAGE}/bootfiles/
+}
+```
+
+We also needed to define the
+[IMAGE_BOOT_FILES](https://docs.yoctoproject.org/singleindex.html#term-IMAGE_BOOT_FILES)
+, which lists files that should
+be installed into the boot partition by the wic tool. We defined it in
+`kas.conf`. Note that the value of this variable heavily depends on your
+system's boot partition layout:
+
+```conf
+IMAGE_BOOT_FILES = " \
+  bootfiles/EFI/BOOT/grub*;EFI/BOOT/ \
+  bootfiles/EFI/BOOT/x86_64-efi/*;EFI/BOOT/x86_64-efi/ \
+  bzImage-initramfs-${MACHINE}.bin;bzImage-initramfs-a \
+  bzImage-initramfs-${MACHINE}.bin;bzImage-initramfs-b \
+  bzImage-initramfs-${MACHINE}.bin.sig;bzImage-initramfs-a.sig \
+  bzImage-initramfs-${MACHINE}.bin.sig;bzImage-initramfs-b.sig \
+  bootx64.efi;EFI/BOOT/ \
```

Comment:
```suggestion
IMAGE_BOOT_FILES = " \
  bootfiles/EFI/BOOT/grubx64.efi;EFI/BOOT/bootx64.efi \
  bootfiles/EFI/BOOT/grubenv*;EFI/BOOT/
  bootfiles/EFI/BOOT/grub.cfg*;EFI/BOOT/
  bootfiles/EFI/BOOT/x86_64-efi/*;EFI/BOOT/x86_64-efi/ \
  bzImage-initramfs-${MACHINE}.bin;bzImage-initramfs \
  bzImage-initramfs-${MACHINE}.bin.sig;bzImage-initramfs.sig \
```

With `MOK_SB = 0` this will be the correct entry for IMAGE_BOOT_FILES; also there is no need to mention hera about bzImage-a and bzImage-b; we do not write here about update system.